R/bcf.R

```diff
@@ -281,161 +301,198 @@ bcf <- function(y, z, x_control, x_moderate=x_control, pihat, w = NULL,
-#' @param nchains numeric, number of chains to launch in parallel (for convergence diagnostics). Should be less or equal to number of cores on the machine.
-#' @return A list with elements
-#' \item{tau}{\code{nsim} by \code{n} matrix of posterior samples of individual treatment effects}
-#' \item{mu}{\code{nsim} by \code{n} matrix of posterior samples of individual treatment effects}
-#' \item{sigma}{Length \code{nsim} vector of posterior samples of sigma}
-#' @import magrittr foreach
-#' @importFrom coda gelman.diag 
-#' @importFrom purrr map
-#' @importFrom parallel makeCluster
-#' @importFrom doParallel registerDoParallel
-#' @export
-parbcf <- function(y, z, x_control, x_moderate=x_control, pihat, w = NULL, 
-                n_threads = max(RcppParallel::defaultNumThreads()/2,1),
-                random_seed = c(1, 235, 2134, 1236),
-                nburn, nsim, nthin = 1, update_interval = 100,
-                ntree_control = 200,
-                sd_control = 2*sd(y),
-                base_control = 0.95,
-                power_control = 2,
-                ntree_moderate = 50,
-                sd_moderate = sd(y),
-                base_moderate = 0.25,
-                power_moderate = 3,
-                nu = 3, lambda = NULL, sigq = .9, sighat = NULL,
-                include_pi = "control", use_muscale=TRUE, use_tauscale=TRUE, verbose=FALSE,
-                nchains = 4
-) {
+    Tm = fitbcf$b_post[,order(perm)] * (1.0/ (fitbcf$b1 - fitbcf$b0))
 
- if(length(random_seed) != nchains){
-    stop("you must provide a random seed for each chain (default 4)")
-    }
+    Tc = ac * (1.0/fitbcf$msd) 
+
+    tau_post = sdy*fitbcf$b_post[,order(perm)]
 
-  cl <- parallel::makeCluster(nchains)
-  doParallel::registerDoParallel(cl)
-
-  fits <- foreach::foreach(i = 1:nchains) %dopar% {
-    library(bcf2)
-
-    bcffit <- bcf(y, z, x_control, x_moderate=x_control, pihat, w, 
-                n_threads,
-                random_seed,
-                nburn, nsim, nthin, update_interval,
-                ntree_control,
-                sd_control,
-                base_control,
-                power_control,
-                ntree_moderate,
-                sd_moderate,
-                base_moderate,
-                power_moderate,
-                nu, lambda, sigq, sighat,
-                include_pi, use_muscale, use_tauscale, verbose
-) 
+    mu_post  = muy + sdy*(Tc*fitbcf$msd + Tm*fitbcf$b0)
+    
+    list(sigma = sdy*fitbcf$sigma,
+         yhat = muy + sdy*fitbcf$yhat_post[,order(perm)],
+         sdy = sdy,
+         con_sd = con_sd,
+         mod_sd = mod_sd,
+         muy = muy,
+         mu  = mu_post,
+         tau = tau_post,
+         mu_scale = fitbcf$msd,
+         tau_scale = fitbcf$bsd,
+         b0 = fitbcf$b0,
+         b1 = fitbcf$b1,
+         perm = perm,
+         include_pi = include_pi,
+         random_seed=this_seed
+    )
+    
   }
+
```

Comment:
@cdelannoy  This section below here is something I particularly want review on, it feels like awkward code to me and I wonder if there's a better way to do it.

Reply:
I looked into this, and found other ways to do it in base R, but they are equally or more clumsy than your method. So I suggest that we keep it the way you wrote it.